st2actions/st2actions/utils/param_utils.py

```diff
@@ -45,295 +43,158 @@ def pf(params, skips):
-        # No override if param is immutable
-        if param_value.get('immutable', False):
-            continue
+def _cast(v, parameter_schema):
+    if v is None or not parameter_schema:
+        return v
 
-        # Check if param exists in action_parameters and if it has a default value then
-        # pickup the override.
-        if param_name in action_parameters:
-            action_param = action_parameters[param_name]
-            if 'default' in action_param:
-                resolved_params[param_name] = action_param['default']
+    parameter_type = parameter_schema.get('type', None)
+    if not parameter_type:
+        return v
 
-            # No further override (from liveaction) if param is immutable
-            if action_param.get('immutable', False):
-                continue
+    cast = get_cast(cast_type=parameter_type)
+    if not cast:
+        return v
 
-        # Finally pick up override from actionexec_runner_parameters
-        if param_name in actionexec_runner_parameters:
-            resolved_params[param_name] = actionexec_runner_parameters[param_name]
+    return cast(v)
 
-    return resolved_params
 
+def process(G, name, value):
+    env = Environment(undefined=StrictUndefined)
+    if isinstance(value, str):
+        value = to_unicode(value)
+    template_ast = env.parse(value)
+    dependencies = meta.find_undeclared_variables(template_ast)
+    if dependencies:
+        G.add_node(name, template=value)
+        for dependency in dependencies:
+            G.add_edge(dependency, name)
+    else:
+        G.add_node(name, value=value)
+
+def render_live_params(runnertype_parameter_info, action_parameter_info, liveaction_parameters,
+                       action_context):
+    '''
+    This function would happen on a st2api side and would only apply to manual action invocations
+    and calls from mistral
+    '''
+    render_context = {}
+    env = Environment(undefined=StrictUndefined)
 
-def _get_resolved_action_params(runner_parameters, action_parameters,
-                                actionexec_action_parameters):
-    # Create action parameters by merging default values with dynamic values
-    resolved_params = {k: v['default'] if 'default' in v else None
-                       for k, v in six.iteritems(action_parameters)
-                       if k not in runner_parameters}
+    G = nx.DiGraph()
+    G.add_node(SYSTEM_KV_PREFIX, value=KeyValueLookup())
+    G.add_node(ACTION_CONTEXT_KV_PREFIX, value=action_context)
 
-    # pick overrides from actionexec_action_parameters
-    for param_name, param_value in six.iteritems(action_parameters):
-        # No override if param is immutable
-        if param_value.get('immutable', False):
-            continue
-        if param_name in actionexec_action_parameters and param_name not in runner_parameters:
-            resolved_params[param_name] = actionexec_action_parameters[param_name]
+    for name, value in six.iteritems(liveaction_parameters):
```

Comment:
The order we add them in the graph not important in this case. Until we start traversing it, the graph is just a pile of nodes and edges. Nodes we create manually with `add_node` would have either template or value property and the nodes that are created automatically as one of the points in `add_edge` would be empty. If we end up with an empty node, it would mean a missing dependency.